### PR TITLE
Add rcon client

### PR DESCRIPTION
**Pull request by tripplet**
This adds a minimal rcon client to allow general script usage.
Additionally there is a players-online.sh script to help with automated updates using watchtower.

### CODE REVIEW

**Comment by Fank** [comment]
I really love the idea, the only thing which prevents me from merging it right away is that I don't like adding the rcon code in this repository directly.
Because its starting to add more and more stuff which is not directly related to docker and factory.
@tripplet Where did you got the rcon.c from? did you wrote it yourself?
I was thinking about moving it a dedicated repository and build binaries there, then load the build binary into this image.
So that everyone who wants to use the binary directly can just use it without loading the whole docker image.

**Comment by tripplet** [comment]
> @tripplet Where did you got the rcon.c from? did you wrote it yourself?

Yes I wrote it myself for exactly this purpose. It's mostly the default/generic code to open a tcp socket in linux + some very basic [RCON](https://developer.valvesoftware.com/wiki/Source_RCON_Protocol).

> So that everyone who wants to use the binary directly can just use it without loading the whole docker image.

Yes that would make sense but I someone is searching for a "fully featured" rcon client they are better of with the many other implementations out there.
If you look very closely it even skips freeing the `char* password` or the result of `combine_args()`.
But as this is freed up anyway by the OS after exit see I see in this case no point for such a simple "one-shot" binary.
It's really written for this exact purpose.
All required variables are even pulled from the factorio images environment variables, there is no way to set a different host, port or password.

It can be compiled with just the following command, the `Makefile` is just there to make it cleaner:
```bash
gcc main.c -o rcon
```
The result binary is just 34kB with no external dependencies from the ~200 lines of C code.

> Because its starting to add more and more stuff which is not directly related to docker and factory.

Yeah I can understand that, my argument would be to provide a working RCON implementation out of the box with the factorio image to not burden the user of the docker image to then search for: Which RCON client do I use for basic usage.

Also adding it to the image makes it easily possible to use RCON without even exposing the port.
(This could of course also be achieved with compose file and a sidecar container).

> I was thinking about moving it a dedicated repository and build binaries there, then load the build binary into this image.

The problem would be that this then would have to mirror the arm64/amd64 compatibility and all the build complexity.
If the C code is just in this repository I there is the need to change something its a simple change to the C file, rebuild the docker image => done.

I think using the [multi-stage build](https://docs.docker.com/build/building/multi-stage/) in the Dockerfile even avoids complexity because the effect on the result container is only the addition of a single 34kB binary.

I hope I have given some good points for your concern to add this complexity but I can of course understand your approach of keeping this repository minimal.

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,13 @@
-FROM debian:stable-slim
+FROM debian:stable-slim AS builder
```
Comment (Fank) [inline docker/Dockerfile:1]:
```suggestion
# build rcon client
FROM debian:stable-slim AS rcon-builder
```

File: docker/Dockerfile [hunk 2]
```diff
@@ -81,6 +89,7 @@ RUN set -ox pipefail \
 
 COPY files/*.sh /
 COPY files/config.ini /opt/factorio/config/config.ini
+COPY --from=builder /src/rcon /bin/rcon
```
Comment (Fank) [inline docker/Dockerfile:92]:
```suggestion
COPY --from=rcon-builder /src/rcon /bin/rcon
```

**Review by Fank — CHANGES_REQUESTED** [review]

**Comment by Fank** [comment]
How about adding `LABEL`s for watchtower using `players-online.sh` in the Dockerfile or mentioning them in the README?
like `LABEL com.centurylinklabs.watchtower.lifecycle.pre-check="/players-online.sh"`

**Comment by tripplet** [comment]
> How about adding `LABEL`s for watchtower using `players-online.sh` in the Dockerfile or mentioning them in the README? like `LABEL com.centurylinklabs.watchtower.lifecycle.pre-check="/players-online.sh"`

Hmm good question, the other option would be to suggestion it in the example compose file and mentioned it in the README rather than in the Dockerfile because I think it's not possible to remove or overwrite a label which is set directly on the container. 
(I haven't tested that but the [internet](https://stackoverflow.com/questions/58062007/docker-compose-can-i-update-traefik-labels-without-restarting-a-container#:~:text=At%20a%20Docker%20API%20level,container%20to%20change%20its%20labels.) seems to suggest that it is no possible to overwrite a label but that's from 2019 so it might no longer be the case)

So it adding it to the Dockerfile would prevent a user from just going yolo I always want the latest container don't care if people are currently playing 😄 

Here is for example my current setup extended by the new pre-check. (Not tested yet).
Also it might be better to use pre-update to have watchtower check for update also if players are online but not do the update?

```yaml
services:
  factorio:
    container_name: factorio
    image: factoriotools/factorio:stable
    restart: unless-stopped
    ports:
      - "34197:34197/udp"
      - "27015:27015/tcp"
    volumes:
      - ./data:/factorio
    environment:
      - CONSOLE_LOG_LOCATION=/factorio/log
      - UPDATE_MODS_ON_START=true
    labels:
      # Allow autoupdate from watchtower only if no players are online
      - com.centurylinklabs.watchtower.enable=true
      - com.centurylinklabs.watchtower.scope=factorio
      - com.centurylinklabs.watchtower.lifecycle.pre-check="/players-online.sh"

  watchtower:
    container_name: watchtower_factorio
    image: containrrr/watchtower
    restart: unless-stopped
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      # Only update containers which have the option 'watchtower.enable=true' set
      - WATCHTOWER_TIMEOUT=30s
      - WATCHTOWER_LABEL_ENABLE=true
      - WATCHTOWER_POLL_INTERVAL=3600
      - WATCHTOWER_LIFECYCLE_HOOKS=true
    labels:
      - com.centurylinklabs.watchtower.scope=factorio
    env_file:
      # Notification on container update 
      # see: https://containrrr.dev/watchtower/notifications/ and https://containrrr.dev/shoutrrr
      - path: ./watchtower_notification.env
        required: false
```

**Comment by Fank** [comment]
@tripplet How about adding a docker-compose.yml to the root folder of the respository with watchtower and the labels added but commented out?
So everyone can run the default image and add the version update check if they want to.
This would make a good solution for everyone i guess.

And the docker-compose.yml in the docker folder is for building the image as all other files are too.

**Comment by Kariton** [comment]
while this looks promising, just not for me though. :)
i just want to drop that link:

https://github.com/nekomeowww/factorio-rcon-api

**Comment by tripplet** [comment]
> How about adding a docker-compose.yml to the root folder of the respository with watchtower and the labels added but commented out?

Good idea I just pushed add an example file.

> this looks promising, just not for me though. :)
> i just want to drop that link:...

Ah very cool project thanks for that Im might use that in the future.

Some other off topic:
I also have another project which is not fully ready for release (still some bugs to solve).
It's another side-car container which just watches the log file for login/logouts and is able to send notifications out via Telegram.
I found this helpful for a server used by a small number of friends as a low effort way to play together.
The idea is that you don't have to write in a group every time to say: I want to play does someone want to join? 
All involved just receive an automatic notification when someone else joins the server and can decide if they want to join or ignore it.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      - "27015:27015/tcp"
+    volumes:
+      - ./data:/factorio
+    environment:
+      - UPDATE_MODS_ON_START=true
+    #labels:
+    #  # Labels to allow autoupdate only if no players are online
+    #  - com.centurylinklabs.watchtower.enable=true
+    #  - com.centurylinklabs.watchtower.scope=factorio
+    #  - com.centurylinklabs.watchtower.lifecycle.pre-update="/players-online.sh"
```
Comment (Fank) [inline docker-compose.yml:18]:
```suggestion
    
    # Uncomment to enable autoupdate via watchtower
    #labels:
    #  # Labels to allow watchtower autoupdate only if no players are online
    #  - com.centurylinklabs.watchtower.enable=true
    #  - com.centurylinklabs.watchtower.scope=factorio
    #  - com.centurylinklabs.watchtower.lifecycle.pre-update="/players-online.sh"
```

**Review by Fank — CHANGES_REQUESTED** [review]

**Comment by Fank** [comment]
Just one more, and i think we are ready to go.

File: README.md [hunk 4]
```diff
@@ -96,6 +96,15 @@ docker run -d -it  \
 docker attach factorio
 ```
 
+### RCON
```
Comment (Fank) [inline README.md:99]:
```suggestion
### RCON (2.0.18+)
```

**Comment by tripplet** [comment]
Awesome, thanks for merging 😃

**Pull request merged**